### ISSUE TITLE
No simple way to store the outputs of a network

### ISSUE THREAD

**Issue by jfsantos** [pluskid/Mocha.jl#59]
I am working on a network that processes audio files (in a frame-by-frame basis) and would like to listen to the outputs. However, there is no easy way to get a snapshot of a model and extract the output of a `forward` pass for a single sample (or for a batch of samples). I am going to use something similar to this [gist](https://gist.github.com/jfsantos/ca147bef179a23911bd3), but that is overly complicated and has to be done for each network separately. 

How and what could we implement something to make this task easier?

**Comment by pluskid** [comment]
@jfsantos Currently maybe the easiest way is to use an HDF5OutputLayer, the output will be written to a HDF5 file. One option might be to implement a MemoryOutputLayer, which just append the blob from each batch to a huge memory array or a list of memory arrays.

Do you have any idea of any form of interface that could make this kind of life easier?

I was also thinking about providing some common high-level wrapper for models like MLP, etc. so that you do not need to write layer definitions every time if you do not care about customizing every details of every layer. But I have not been able to sit down and working on that yet.

**Comment by jfsantos** [comment]
Thanks, for some reason I did not find the HDF5OutputLayer, but that's still a bit troublesome if you have one program that trains the model and another to do on-the-fly predictions based on that model.

All the deep learning frameworks I'm familiar with (Theano, pylearn2 and Caffe) have a bit of trouble at this part because they are implemented thinking on experiments and exploration, not deployment. Because of that, there's no interface that we could easily "imitate". It would be nice to be able to have a simple function like `predict(mdl::Net, x::Array{T,N})` that gets a net and one or more inputs and generates the corresponding outputs.

Having a high-level wrapper for "straightforward" models is an interesting idea. [Blocks](https://github.com/bartvm/blocks) (a framework based on Theano) does this and it's really useful. The API is like this: `mlp = MLP(activations=[Rectifier(), Softmax()], dims=[784, 100, 10])`. Maybe we could have something similar (with an optional argument to add Dropout layers)? As soon as I finish my current experiments I might be able to try to come up with something for Mocha.

**Comment by mrtzh** [comment]
I also find this issue important. A `predict` method would be very helpful. I know how to write it myself for the CPUBackend (though it isn't pretty), but I'm currently stuck getting it to work for the GPUBackend trying to convert the CuBlas blobs to julia arrays. If someone could briefly comment how best to implement such a `predict` function in both backends, that would really help.

Thanks!

**Comment by pluskid** [comment]
@mrtzh Sorry for delaying for implementing a higher level interface. Concerning your particular function, there is a generic function `to_array(blob)` that converts any blob to a Julia Array. It should work with both CPU blob and GPU Blob.

**Comment by mrtzh** [comment]
Thanks! Now, suppose I just want to extract the classification accuracy into a julia variable. After training the accuracylayer tells me that the `accuracy` attribute is 0 and I have to apply `forward` a bunch of times to compute it. Is there an easier way to just return the accuracy value after training?

**Comment by pluskid** [comment]
@mrtzh Do you want training accuracy or test accuracy? For training accuracy, you can actually have all the training statistics saved to a file and load for plotting. The code for training stacked auto encoder has a demo on how you could save and make plots like this: http://mochajl.readthedocs.org/en/latest/tutorial/mnist-sDA.html

**Comment by mrtzh** [comment]
Both. I'd really just like to be able to train a net and have these numbers written to a julia variable. The tutorial shows how to save them to a jld file and how to load them back in. I find that pretty cumbersome. Currently, I'm computing the the accuracy after training by forwarding the net a bunch of times and computing the accuracy statistics myself. But it just feels like I'm doing it wrong and there should be simple way to do it. If there isn't, that's fine and I'll work with my code. I just want to avoid reinventing the wheel.

**Comment by pluskid** [comment]
@mrtzh Sorry, you are right, there is not a standard way to do it right now. :/

**Comment by mrtzh** [comment]
No worries :-) I really admire the work you put into Mocha.jl.

Here's the workaround I'm currently using in case it's helpful to anyone reading this:
```
function predict(net::Net,source_file::String,batch_size=100)
    # select all but first and last layer from net
    common_layers = net.layers[2:end-1]
    data_layer = HDF5DataLayer(name="predict-data", source=source_file, batch_size=batch_size)
    acc_layer = AccuracyLayer(name="predict-accuracy", bottoms=[:ip2, :label])
    predict_net = Net("Predictions",net.backend, [data_layer,common_layers...,acc_layer])
    
    # extract predictions from new net
    predictions = Float64[]
    epoch = get_epoch(predict_net)
    while get_epoch(predict_net) == 0
        forward(predict_net)
        batch = to_array(predict_net.states[end-1].blobs[1])
        for i = 1:batch_size
            push!(predictions,indmax(batch[:,i])-1)
        end
    end
    predictions
end
```
It reads data from disk and applies `net` to each of the data points. To get the `predict` method that @jfsantos was interested in, the following should work:
```
function predict(net::Net,X::Array{Float32,4},batch_size=size(X)[4])
    # select all but first and last layer from net
    common_layers = net.layers[2:end-1]
    data_layer = MemoryDataLayer(name="predict-data", data=Array[X,Array(Float32,1,size(X)[4])], batch_size=batch_size)
    acc_layer = AccuracyLayer(name="predict-accuracy", bottoms=[:ip2, :label])
    predict_net = Net("Predictions",net.backend, [data_layer,common_layers...,acc_layer])
    
    # extract predictions from new net
    predictions = Float64[]
    epoch = get_epoch(predict_net)
    while get_epoch(predict_net) == 0
        forward(predict_net)
        batch = to_array(predict_net.states[end-1].blobs[1])
        for i = 1:batch_size
            push!(predictions,indmax(batch[:,i])-1)
        end
    end
    predictions
end
```
It's similar to his gist, but it used `to_array` so that it also works for `GPUBackend`.

Let me know if this is something worth adding to the examples and I can send a pull request.

**Comment by pluskid** [comment]
Now we have a `forward_epoch` function for the net, and also a `MemoryOutputLayer`, which should be able to simplify the snippet above a bit.
